### PR TITLE
JDK-8281001 Class::forName(String) defaults to system class loader if the caller is null

### PR DESCRIPTION
**Pull request by tprinzing**
The Class::forName behavior change to match JNI FindClass is a compatible change and seems pretty attractive as it would be expected that Class::forName would give the same behavior as FindClass which uses the system classloader.  The test for 8281006 was enhanced to test for this change.  Merged master to pick up fixes to unrelated test failures to reduce noise.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8281001](https://bugs.openjdk.org/browse/JDK-8281001): Class::forName(String) defaults to system class loader if the caller is null
 * [JDK-8286927](https://bugs.openjdk.org/browse/JDK-8286927): Class::forName(String) defaults to system class loader if the caller is null (**CSR**)

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8711/head:pull/8711` \
`$ git checkout pull/8711`

Update a local copy of the PR: \
`$ git checkout pull/8711` \
`$ git pull https://git.openjdk.java.net/jdk pull/8711/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8711`

View PR using the GUI difftool: \
`$ git pr show -t 8711`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8711.diff">https://git.openjdk.java.net/jdk/pull/8711.diff</a>

</details>

### CODE REVIEW

**Comment by jddarcy** [comment]
@tprinzing , please file a CSR for this discuss so the behavioral changes can be reviewed.

File: test/jdk/java/lang/module/exeNullCallerGetResource/NullCallerGetResource.java [hunk 1]
```diff
@@ -25,6 +25,7 @@
 /**
  * @test
  * @bug 8281006
+ * @bug 8281001
```
Comment (mlchung) [inline test/jdk/java/lang/module/exeNullCallerGetResource/NullCallerGetResource.java:28]:
```suggestion
 * @bug 8281006 8281001
```

`@bug` can have one or more bug IDs

File: src/java.base/share/classes/java/lang/Class.java [hunk 2]
```diff
@@ -380,7 +380,10 @@ public static Class<?> forName(String className)
     @CallerSensitiveAdapter
     private static Class<?> forName(String className, Class<?> caller)
             throws ClassNotFoundException {
-        return forName0(className, true, ClassLoader.getClassLoader(caller), caller);
+        ClassLoader loader = (caller == null) ?
+                ClassLoader.getSystemClassLoader() :
+                ClassLoader.getClassLoader(caller);
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/Class.java:385]: Formatting nit:
```
   ClassLoader loader = (caller == null) ? ClassLoader.getSystemClassLoader()
                                         : ClassLoader.getClassLoader(caller);
```

**Review by mlchung — COMMENTED** [review]
`Class::forName(String)` javadoc should specify which class loader to use when invoked with null caller similar to the other APIs you fixed for null callers.

I think this new test case does not belong to `NullCallerGetResource.java` which is a test for `Module::getResource`.    It's better to be placed under the `test/jdk/java/lang/Class/forName` directory that makes it clear it's a test for `Class::forName`. 

Alternatively, we could move all the tests for null caller under a new clearly-named directory, maybe `test/jdk/jdk/nullCaller`.    This may allow to do some refactoring and remove the duplicated code in these test cases.   What do you think?

**Comment by iklam** [comment]
Please rename the title of the issue to reflect what is being proposed.

**Comment by tprinzing** [comment]
> `Class::forName(String)` javadoc should specify which class loader to use when invoked with null caller similar to the other APIs you fixed for null callers.
> 
> I think this new test case does not belong to `NullCallerGetResource.java` which is a test for `Module::getResource`. It's better to be placed under the `test/jdk/java/lang/Class/forName` directory that makes it clear it's a test for `Class::forName`.
> 
> Alternatively, we could move all the tests for null caller under a new clearly-named directory, maybe `test/jdk/jdk/nullCaller`. This may allow to do some refactoring and remove the duplicated code in these test cases. What do you think?

I like the idea of moving all the null caller tests to a clearly named directory to avoid duplication.

**Comment by mlchung** [comment]
> I like the idea of moving all the null caller tests to a clearly named directory to avoid duplication.

+1.   You can do this refactoring in a separate JBS issue and then update this PR when the tests are moved.

File: test/jdk/jni/nullCaller/NullCallerTest.java [hunk 3]
```diff
@@ -24,7 +24,7 @@
 
 /**
  * @test
- * @bug 8280902 8281000 8281003 8281006
+ * @bug 8280902 8281000 8281001 8281003 8281006
```
Comment (mlchung) [inline test/jdk/jni/nullCaller/NullCallerTest.java:27]: nit: append the bug rather than keeping the list in an increasing order.

**Review by mlchung — APPROVED** [review]
Looks good.  Thanks for refactoring the tests, making the addition of a new test case much cleaner.

**Review by mlchung — APPROVED** [review]

**Pull request closed without merging**